README.rst

.. image:: https://github.com/sdpython/wrapclib/blob/master/_doc/sphinxdoc/source/phdoc_static/project_ico.png?raw=true
    :target: https://github.com/sdpython/wrapclib/

.. _l-README:

wrapclib
========

.. image:: https://travis-ci.org/sdpython/wrapclib.svg?branch=master
    :target: https://travis-ci.org/sdpython/wrapclib
    :alt: Build status

.. image:: https://ci.appveyor.com/api/projects/status/auonxiihm1eihv3t?svg=true
    :target: https://ci.appveyor.com/project/sdpython/wrapclib
    :alt: Build Status Windows

.. image:: https://circleci.com/gh/sdpython/wrapclib/tree/master.svg?style=svg
    :target: https://circleci.com/gh/sdpython/wrapclib/tree/master

.. image:: https://dev.azure.com/xavierdupre3/wrapclib/_apis/build/status/sdpython.wrapclib
    :target: https://dev.azure.com/xavierdupre3/wrapclib/

.. image:: https://badge.fury.io/py/wrapclib.svg
    :target: https://pypi.org/project/wrapclib/

.. image:: https://img.shields.io/badge/license-MIT-blue.svg
    :alt: MIT License
    :target: http://opensource.org/licenses/MIT

.. image:: https://requires.io/github/sdpython/wrapclib/requirements.svg?branch=master
     :target: https://requires.io/github/sdpython/wrapclib/requirements/?branch=master
     :alt: Requirements Status

.. image:: https://codecov.io/github/sdpython/wrapclib/coverage.svg?branch=master
    :target: https://codecov.io/github/sdpython/wrapclib?branch=master

.. image:: http://img.shields.io/github/issues/sdpython/wrapclib.png
    :alt: GitHub Issues
    :target: https://github.com/sdpython/wrapclib/issues

.. image:: http://www.xavierdupre.fr/app/wrapclib/helpsphinx/_images/nbcov.png
    :target: http://www.xavierdupre.fr/app/wrapclib/helpsphinx/all_notebooks_coverage.html
    :alt: Notebook Coverage

.. image:: https://pepy.tech/badge/wrapclib/month
    :target: https://pepy.tech/project/wrapclib/month
    :alt: Downloads

.. image:: https://img.shields.io/github/forks/sdpython/wrapclib.svg
    :target: https://github.com/sdpython/wrapclib/
    :alt: Forks

.. image:: https://img.shields.io/github/stars/sdpython/wrapclib.svg
    :target: https://github.com/sdpython/wrapclib/
    :alt: Stars

.. image:: https://img.shields.io/github/repo-size/sdpython/wrapclib
    :target: https://github.com/sdpython/wrapclib/
    :alt: size

*wrapclib* is a wrapper for some C libraries difficult to build
otherwise. One if them is
`re2 <https://github.com/google/re2>`_
with some taken from
`pyre2 <https://github.com/facebook/pyre2>`_.

::

    from wrapclib import re2
    s = "<h1>mot</h1>"
    print(re2.compile("(<.*>)").match(s).groups())

**Links:**

* `GitHub/wrapclib <https://github.com/sdpython/wrapclib/>`_
* `documentation <http://www.xavierdupre.fr/app/wrapclib/helpsphinx/index.html>`_
* `Blog <http://www.xavierdupre.fr/app/wrapclib/helpsphinx/blog/main_0000.html#ap-main-0>`_


HISTORY.rst

.. _l-HISTORY:

=======
History
=======

current - 2021-01-09 - 0.00Mb
=============================

* `4`: Update submodule re2 to the latest (2021-01-09)

0.1.30 - 2019-09-07 - 0.20Mb
============================

0.1.29 - 2019-09-07 - 0.15Mb
============================

0.1.27 - 2019-09-07 - 0.15Mb
============================

0.1.11 - 2019-03-15 - 0.03Mb
============================

* `2`: add method findall (2019-03-15)
* `1`: build with re2 (2019-03-15)


_unittests/ut_re2/test_matches.py
"""
@brief      test log(time=3s)
"""
import unittest
from pyquickhelper.pycode import ExtTestCase
from wrapclib import re2


class TestMatch(ExtTestCase):

    def test_const_match(self):
        m = re2.match('abc', 'abc')
        self.assertIsNotNone(m)
        self.assertEqual(m.start(), 0)
        self.assertEqual(m.end(), 3)
        self.assertEqual(m.span(), (0, 3))
        self.assertEqual(m.groups(), tuple())
        self.assertEqual(m.groupdict(), {})
        self.assertEqual(m.pos, 0)
        self.assertEqual(m.endpos, 3)
        self.assertEqual(m.string, 'abc')
        self.assertIsNotNone(m.re)
        self.assertEqual(m.re.pattern, 'abc')
        self.assertEqual(m.re.groups, 0)
        self.assertEqual(m.re.groupindex, {})

    def test_group_match(self):
        m = re2.match('ab([cde]fg)', 'abdfghij')
        self.assertIsNotNone(m)
        self.assertEqual(m.start(), 0)
        self.assertEqual(m.end(), 5)
        self.assertEqual(m.span(), (0, 5))
        self.assertEqual(m.groups(), ('dfg',))
        self.assertEqual(m.groupdict(), {})
        self.assertEqual(m.pos, 0)
        self.assertEqual(m.endpos, 8)
        self.assertEqual(m.string, 'abdfghij')
        self.assertIsNotNone(m.re)
        self.assertEqual(m.re.pattern, 'ab([cde]fg)')
        self.assertEqual(m.re.groups, 1)
        self.assertEqual(m.re.groupindex, {})

    def test_named_group_match(self):
        m = re2.match('ab(?P<testgroup>[cde]fg)', 'abdfghij')
        self.assertIsNotNone(m)
        self.assertEqual(m.start(), 0)
        self.assertEqual(m.end(), 5)
        self.assertEqual(m.span(), (0, 5))
        self.assertEqual(m.groups(), ('dfg',))
        self.assertEqual(m.groupdict(), {'testgroup': 'dfg'})
        self.assertEqual(m.pos, 0)
        self.assertEqual(m.endpos, 8)
        self.assertEqual(m.string, 'abdfghij')
        self.assertIsNotNone(m.re)
        self.assertEqual(m.re.pattern, 'ab(?P<testgroup>[cde]fg)')
        self.assertEqual(m.re.groups, 1)
        self.assertEqual(m.re.groupindex, {'testgroup': 1})

    def test_compiled_match(self):
        r = re2.compile('ab([cde]fg)')
        m = r.match('abdfghij')
        self.assertIsNotNone(m)
        self.assertEqual(m.start(), 0)
        self.assertEqual(m.end(), 5)
        self.assertEqual(m.span(), (0, 5))
        self.assertEqual(m.groups(), ('dfg',))
        self.assertEqual(m.groupdict(), {})
        self.assertEqual(m.pos, 0)
        self.assertEqual(m.endpos, 8)
        self.assertEqual(m.string, 'abdfghij')
        self.assertIsNotNone(m.re)
        self.assertEqual(m.re.pattern, 'ab([cde]fg)')
        self.assertEqual(m.re.groups, 1)
        self.assertEqual(m.re.groupindex, {})

    def test_match_raise(self):
        '''test that using the API incorrectly fails'''
        r = re2.compile('ab([cde]fg)')
        self.assertRaise(lambda: re2.match(r, 'abdfghij'), TypeError)

    def test_match_bytes(self):
        ''' test that we can match things in the bytes type '''
        r = re2.compile('(\\x09)')
        m = r.match(b'\x09')
        self.assertIsNotNone(m)
        g = m.groups()
        self.assertTrue(isinstance(g, tuple))
        self.assertTrue(isinstance(g[0], bytes))
        self.assertEqual(b'\x09', g[0])
        self.assertEqual(m.pos, 0)
        self.assertEqual(m.endpos, 1)
        self.assertEqual(m.string, b'\x09')
        self.assertIsNotNone(m.re)
        self.assertEqual(m.re.pattern, '(\\x09)')
        self.assertEqual(m.re.groups, 1)
        self.assertEqual(m.re.groupindex, {})

    def test_match_str(self):
        ''' test that we can match binary things in the str type '''
        r = re2.compile('(\\x09)')
        m = r.match('\x09')
        self.assertIsNotNone(m)
        g = m.groups()
        self.assertTrue(isinstance(g, tuple))
        self.assertTrue(isinstance(g[0], str))
        self.assertEqual('\x09', g[0])

    def test_match_bad_utf8_bytes(self):
        ''' Validate that we just return None on invalid utf-8 '''
        r = re2.compile('\\x80')
        m = r.match(b'\x80')
        self.assertIsNone(m)

    def test_invalid_pattern(self):
        ''' Verify that bad patterns raise an exception '''
        self.assertRaise(lambda: re2.compile(')'), Exception)

    def test_span_type(self):
        ''' verify that start/end return the native literal integer type '''
        r = re2.compile('abc')
        m = r.match('abc')
        self.assertTrue(isinstance(m.start(), type(1)))
        self.assertTrue(isinstance(m.end(), type(1)))

    def test_set_unanchored(self):
        s = re2.Set(re2.UNANCHORED)
        s_with_default_anchoring = re2.Set()

        for re2_set in [s, s_with_default_anchoring]:
            re2_set.add('foo')
            re2_set.add('bar')
            re2_set.add('baz')
            re2_set.compile()

            self.assertEqual(re2_set.match('foo'), [0])
            self.assertEqual(re2_set.match('bar'), [1])
            self.assertEqual(re2_set.match('baz'), [2])
            self.assertEqual(re2_set.match('afoobaryo'), [0, 1])

            self.assertEqual(re2_set.match('ooba'), [])

    def test_set_anchor_both(self):
        s = re2.Set(re2.ANCHOR_BOTH)
        self.assertEqual(s.add('foo'), 0)
        self.assertEqual(s.add('bar'), 1)
        s.compile()

        self.assertEqual(s.match('foobar'), [])
        self.assertEqual(s.match('fooba'), [])
        self.assertEqual(s.match('oobar'), [])
        self.assertEqual(s.match('foo'), [0])
        self.assertEqual(s.match('bar'), [1])

    def test_set_anchor_start(self):
        s = re2.Set(re2.ANCHOR_START)
        self.assertEqual(s.add('foo'), 0)
        self.assertEqual(s.add('bar'), 1)
        s.compile()

        self.assertEqual(s.match('foobar'), [0])
        self.assertEqual(s.match('oobar'), [])
        self.assertEqual(s.match('foo'), [0])
        self.assertEqual(s.match('ofoobaro'), [])
        self.assertEqual(s.match('baro'), [1])

    def test_bad_anchoring(self):
        self.assertRaise(lambda: re2.Set(None), ValueError,
                         'anchoring must be one of')
        self.assertRaise(lambda: re2.Set(15), ValueError,
                         'anchoring must be one of')
        self.assertRaise(lambda: re2.Set({}), ValueError,
                         'anchoring must be one of')

    def test_match_without_compile(self):
        s = re2.Set()
        s.add('foo')
        self.assertRaise(lambda: s.match('bar'),
                         RuntimeError, 'Can\'t match() on an')

    def test_add_after_compile(self):
        s = re2.Set()
        s.add('foo')
        s.compile()
        self.assertRaise(lambda: s.add('bar'), RuntimeError,
                         'Can\'t add() on an already compiled Set')

    def test_double_compile(self):
        s = re2.Set()
        s.add('foo')
        s.compile()
        s.compile()

    def test_add_with_bad_pattern(self):
        s = re2.Set()
        self.assertRaise(lambda: s.add('('), ValueError, 'missing')
        self.assertRaise(lambda: s.add(3), TypeError)

    def test_findall(self):
        s = """date 0 : 14/9/2000 date 1 : 20/04/1971 """
        reg = re2.compile(
            "([0-3]?[0-9]/[0-1]?[0-9]/([0-2][0-9])?[0-9][0-9])[^\\d]")
        res = reg.search(s, 0)
        self.assertEqual(res.groups(), ('14/9/2000', '20'))
        self.assertEqual(res.pos, 0)
        self.assertEqual(res.endpos, 39)
        self.assertEqual(res.span(), (9, 19))
        fall = re2.findall(reg, s)
        self.assertEqual(fall, [('14/9/2000', '20'), ('20/04/1971', '19')])
        fall = re2.findall(
            "([0-3]?[0-9]/[0-1]?[0-9]/([0-2][0-9])?[0-9][0-9])[^\\d]", s)
        self.assertEqual(fall, [('14/9/2000', '20'), ('20/04/1971', '19')])


if __name__ == "__main__":
    unittest.main()


_unittests/ut_re2/test_missing.py
"""
@brief      test log(time=3s)
"""
import unittest
from pyquickhelper.pycode import ExtTestCase
from wrapclib import re2


class TestMissing(ExtTestCase):

    def test_search(self):
        m = re2.search('abc', 'abc')
        self.assertIsNotNone(m)
        self.assertEqual(m.start(), 0)

    def test_search_endpos(self):
        m = re2.search('ab', 'abc', endpos=2)
        self.assertIsNotNone(m)
        self.assertEqual(m.start(), 0)

    def test_fullmatch(self):
        m = re2.fullmatch('abc', 'abc')
        self.assertIsNotNone(m)
        self.assertEqual(m.start(), 0)

    def test_findall(self):
        m = re2.findall('ab', 'abc', endpos=2)
        self.assertIsNotNone(m)


if __name__ == "__main__":
    unittest.main()


_unittests/ut_module/test_readme.py
"""
@brief      test tree node (time=50s)
"""
import os
import unittest
from pyquickhelper.loghelper import fLOG
from pyquickhelper.pycode import get_temp_folder, check_readme_syntax


class TestReadme(unittest.TestCase):

    def test_venv_docutils08_readme(self):
        fLOG(
            __file__,
            self._testMethodName,
            OutputPrint=__name__ == "__main__")

        fold = os.path.dirname(os.path.abspath(__file__))
        readme = os.path.join(fold, "..", "..", "README.rst")
        self.assertTrue(os.path.exists(readme))
        with open(readme, "r", encoding="utf8") as f:
            content = f.read()

        self.assertTrue(len(content) > 0)
        temp = get_temp_folder(__file__, "temp_readme")

        if __name__ != "__main__":
            # does not work from a virtual environment
            return

        check_readme_syntax(readme, folder=temp, fLOG=fLOG)


if __name__ == "__main__":
    unittest.main()


wrapclib/__init__.py
# -*- coding: utf-8 -*-
"""
@file
@brief Module *wrapclib*.
:epkg:`Python` + some C libraries.
"""

__version__ = "0.1.70"
__author__ = "Xavier Dupré"
__github__ = "https://github.com/sdpython/wrapclib"
__url__ = "http://www.xavierdupre.fr/app/wrapclib/helpsphinx/index.html"
__license__ = "MIT License"
__blog__ = """
<?xml version="1.0" encoding="UTF-8"?>
<opml version="1.0">
    <head>
        <title>blog</title>
    </head>
    <body>
        <outline text="wrapclib"
            title="wrapclib"
            type="rss"
            xmlUrl="http://www.xavierdupre.fr/app/wrapclib/helpsphinx/_downloads/rss.xml"
            htmlUrl="http://www.xavierdupre.fr/app/wrapclib/helpsphinx/blog/main_0000.html" />
    </body>
</opml>
"""


def check(log=False):
    """
    Checks the library is working.
    It raises an exception.
    If you want to disable the logs:

    @param      log     if True, display information, otherwise
    @return             0 or exception
    """
    return True


def _setup_hook(use_print=False):
    """
    if this function is added to the module,
    the help automation and unit tests call it first before
    anything goes on as an initialization step.
    """
    # we can check many things, needed module
    # any others things before unit tests are started
    if use_print:
        print("Success: _setup_hook")


wrapclib/re2/re2.py
"""
Copyright (c) 2010, David Reiss and Facebook, Inc. All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:
* Redistributions of source code must retain the above copyright
  notice, this list of conditions and the following disclaimer.
* Redistributions in binary form must reproduce the above copyright
  notice, this list of conditions and the following disclaimer in the
  documentation and/or other materials provided with the distribution.
* Neither the name of Facebook nor the names of its contributors
  may be used to endorse or promote products derived from this software
  without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
"""

import sre_constants
from ._re2 import _compile  # pylint: disable=E0611
from ._re2 import escape, Set, UNANCHORED, ANCHOR_START, ANCHOR_BOTH  # pylint: disable=E0611

__all__ = [
    "error",
    "escape",
    "compile",
    "search",
    "match",
    "findall",
    "fullmatch",
    "Set",
    "UNANCHORED",
    "ANCHOR_START",
    "ANCHOR_BOTH",
]

error = sre_constants.error  # pylint: disable=E1101


def compile(pattern):  # pylint: disable=W0622
    "Compiles a regular expression pattern, returning a pattern object."
    return _compile(pattern, error)


def search(pattern, string, pos=0, endpos=-1):
    """Scans through string looking for a match to the pattern, returning
    a match object, or None if no match was found."""
    if endpos != -1:
        return _compile(pattern, error).search(string, pos, endpos)
    else:
        return _compile(pattern, error).search(string, pos)


def match(pattern, string):
    """Tries to apply the pattern at the start of the string, returning
    a match object, or None if no match was found."""
    return _compile(pattern, error).match(string)


def fullmatch(pattern, string):
    """Tries to apply the pattern to the entire string, returning
    a match object, or None if no match was found."""
    return _compile(pattern, error).fullmatch(string)


def findall(pattern, string, pos=0, endpos=-1):
    """
    Implements method :epkg:`*py:re:findall`
    for *re2* in :epkg:`Python`.

    @param      pattern     compiled regular expression
    @param      string      string to search
    @param      pos         first position to look into
    @param      endpos      last position to look into (-1 for the last one)
    @return                 list of results

    .. exref::
        :title: Example of findall

        A quick example with method
        @see fn findall.

        .. runpython::
            :showcode:

            from wrapclib import re2

            s = "date 0 : 14/9/2000 date 1 : 20/04/1971 "

            reg = re2.compile(
                "([0-3]?[0-9]/[0-1]?[0-9]/([0-2][0-9])?[0-9][0-9])[^\\d]")

            fall = re2.findall(reg, s)
            print(fall)
    """
    if isinstance(pattern, str):
        pattern = _compile(pattern, error)
    results = []
    if endpos == -1:
        def fsearch(s, p):
            return pattern.search(s, p)
    else:
        def fsearch(s, p):
            return pattern.search(s, p, endpos)
    res = fsearch(string, pos)
    while res:
        results.append(res.groups())
        pos = res.span()[1]
        res = fsearch(string, pos)
    return results
